aiohttp/web_urldispatcher.py

```diff
@@ -230,8 +231,24 @@ def set_current_app(self, app):
     def freeze(self):
         self._frozen = True
 
+    @property
+    def variable_maps(self):
+        return tuple(reversed(self._variables.maps))
+
+    def add_variables(self, match_dict):
+        self._variables.maps.append(match_dict)
```

Comment:
I suspect you should insert a new dict *in front* of maps -- like we do it for `add_app()`.
Moreover maybe merging `add_app()` and `add_variables()` functionality makes sense.
What do you think about `_add_app(app, match_dict)` method as replacement for `add_app`.
Adding underscore emphasizes that the method is private. `add_app()` was never a part of documented public API, I think the change is safe in backward compatibility perspective.

Reply:
`ChainMap` resolves from the beginning of its `maps`. The point here is that `add_variable()` is called *after* resolving the subapp's routes, which means that subapp's match variables are added to the chain map first. Since we want to give priority to the innermost subapp's variables in the merged view of this chain map, this `append` is natural.

I think merging this method into `add_app(app, match_dict)` is a good idea. :+1:

Reply:
Hm.... merging into a single method causes a lot of problem:
* When instantiating `UrlMappingMatchInfo` we already have `match_dict` (so the chain map begins with a single layer) but `_app` is None and `_apps` is an empty tuple. This makes the number of nested maps and apps inconsistent.
* Changing `UrlMappingMatchInfo`'s constructor to accept an explicit `app` argument requires `Route` and `Resource` classes to keep track of their owner `app` because `resolve()` method should be able to indicate the exact `app` instance to create the match info object. `request.app` is a self null reference there.

The root cause of this mess is that `app` and `match_dict` are added to `UrlMappingMatchInfo` at different places and timings. :confused: